helix-core/src/main/java/org/apache/helix/controller/rebalancer/waged/constraints/SoftConstraint.java

```diff
@@ -27,23 +27,31 @@
  * Evaluate a partition allocation proposal and return a score within the normalized range.
  * A higher score means the proposal is more preferred.
  */
-public interface SoftConstraint {
-  float MIN_SCORE = -1000.0f;
-  float MAX_SCORE = 1000.0f;
+public abstract class SoftConstraint {
+  private Type _type;
 
-  /**
-   * The scoring function returns a score between MINIMAL_SCORE and MAXIMUM_SCORE, which is then weighted by the
-   * individual normalized constraint weights.
-   * Each individual constraint will define the meaning of MINIMAL_SCORE to MAXIMUM_SCORE differently.
-   */
-  float assignmentScore(AssignableNode node, AssignableReplica rep, ClusterContext clusterContext);
+  public SoftConstraint(Type type) {
+    _type = type;
+  }
+
+  enum Type {
```

Comment:
If we have 2 constraints with the same type defined, how to separately weight them?

Reply:
Type and constraint need to be 1 to 1 mapping. How can we have one type mapping to multiple constraints?

Reply:
I agree with 1:1 mapping. But this implementation is not protecting this. If it has to be one one mapping, shall we just take the Constraint class as the type?

Reply:
If we use the class object then I guess the enum type definition is meaningless.

Reply:
Yeah, that's my question here. Do we really need the enum here, or not?

Reply:
We can go that route (no enum) but the enum was created in the first place, I do think it's cleaner. Back to the original simple interface design, we could also get away with enums as well since it doesn't also enforce 1:1 mapping

Then the weight map would consist of 
```
Map<SoftConstraint, Float> weights = 
 ImmutableMap.of(LeastUsedNode.class, 1.0f);
```
or 
```
Map<SoftConstraint, Float> weights = 
 ImmutableMap.of(instance of LeastUsedNode.class, 1.0f);
```